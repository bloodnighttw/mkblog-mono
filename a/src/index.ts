import path from "node:path";
import { type PluggableList, unified } from "unified";
import remark2rehype from "remark-rehype";
import remarkParse from "remark-parse";
import rehypeStringify from "rehype-stringify";
import fs from "node:fs";
import type { Root as HastRoot } from "hast";
import type { Root as MDRoot } from "mdast";
import frontmatter from "remark-frontmatter";
import { select } from "unist-util-select";
import yaml from "yaml";
import { z, type ZodRawShape } from "zod";

interface ContentOptions<T extends ZodRawShape> {
	root: string;
	matchRegex?: RegExp | "markdown";
	createSlug?: (from: string) => string;
	schema: T;

	mkBlogPlugins?: () => void;
	remarkPlugins?: PluggableList;
	rehypePlugins?: PluggableList;
}

interface YAMLNode {
	type: "yaml";
	value: string;
}

class Content<T> {
	readonly filepath: string;
	readonly slug: string;
	private readonly remarkPlugins: PluggableList;
	private readonly rehypePlugins: PluggableList;
	private readonly schema: z.ZodSchema<T>;

	constructor(
		filepath: string,
		slug: string,
		remarkPlugins: PluggableList,
		rehypePlugins: PluggableList,
		schema: z.Schema,
	) {
		this.filepath = filepath;
		this.slug = slug;
		this.remarkPlugins = remarkPlugins;
		this.rehypePlugins = rehypePlugins;
		this.schema = schema;
	}

	protected remarkProcessor() {
		return unified()
			.use(remarkParse)
			.use(frontmatter) // for metadata decode
			.use(this.remarkPlugins);
	}

	protected rehypeProcessor() {
		return unified().use(remark2rehype).use(this.rehypePlugins);
	}

	protected rehype2html() {
		return unified().use(rehypeStringify);
	}

	/**
	 * @protected to fetch the markdown and html ast, used internal.
	 */

	protected async mdast(): Promise<MDRoot> {
		const fileContent = await fs.promises.readFile(this.filepath, "utf-8");
		return this.remarkProcessor().parse(fileContent);
	}

	protected async hast(mdroot: MDRoot | null = null): Promise<HastRoot> {
		const temp = mdroot ?? (await this.mdast());
		return await this.rehypeProcessor().run(temp);
	}

	/**
	 *
	 * @param root the hast root, if it's null, we will fetch hast automatically
	 * @protected this method should be called internally
	 * @Return return raw html string of content.
	 */
	protected async html(root: HastRoot | null = null) {
		const temp = root ?? (await this.hast());
		return this.rehype2html().stringify(temp);
	}

	public async metadata(mdast: MDRoot | null = null): Promise<T> {
		const temp = mdast ?? (await this.mdast());
		const yamlContent = select("yaml", temp) as YAMLNode;

		if (yamlContent) {
			const metadata = yaml.parse(yamlContent.value);
			return this.schema.parse(metadata);
		}

		throw new Error("metadata not found");
	}
}

interface Collection<T> {
	posts: Content<T>[];
}


const markdownRegex = /\.md$/;

async function defineContent<T extends ZodRawShape>(
	options: ContentOptions<T>,
): Promise<Collection<T>> {
	const base = path.join(process.cwd(), options.root);
	const changeRegex =
		options.matchRegex === "markdown" || !options.matchRegex
			? markdownRegex
			: options.matchRegex;
	const files = (await fs.promises.readdir(base)).filter((file) =>
		changeRegex.test(file),
	);

	const posts = files.map((file) => {
		const filepath = path.join(base, file);
		const slug =
			options.createSlug?.(file) ?? file.replace(changeRegex, "");
		return new Content<T>(
			filepath,
			slug,
			options.remarkPlugins ?? [],
			options.rehypePlugins ?? [],
			z.object(options.schema),
		);
	});

	return { posts };
}
export type { ContentOptions, Collection };

export { defineContent, Content, z as zod };
