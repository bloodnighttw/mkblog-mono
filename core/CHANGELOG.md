# @mkblog/core

## 1.0.0

### Major Changes

- 0ff20e7: # A Major Version of mkblog.

## 0.4.2

### Patch Changes

- fix i upload the whole package.

## 0.4.1

### Patch Changes

- add type check to project when building

## 0.4.0

### Minor Changes

- use include pattern instead of regex.

## 0.3.1

### Patch Changes

- e7a6a32: Add monorepo and testing.
- 95e9f68: add license
