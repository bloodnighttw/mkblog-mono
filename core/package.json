{
	"name": "@mkblog/core",
	"version": "0.3.1",
	"type": "module",
	"description": "The core library of mkblog",
	"main": "./dist/index.js",
	"types": "./dist/index.d.ts",
	"scripts": {
		"build": "tsup",
		"test": "echo \"Error: no test specified\""
	},
	"dependencies": {
		"fast-glob": "^3.3.3",
		"rehype-stringify": "^10.0.1",
		"remark-frontmatter": "^5.0.0",
		"remark-parse": "^11.0.0",
		"remark-rehype": "^11.1.1",
		"unified": "^11.0.5",
		"unist-util-select": "^5.1.0",
		"yaml": "^2.7.0",
		"zod": "^3.24.2"
	},
	"devDependencies": {
		"@vitest/coverage-v8": "3.0.7",
		"remark-gfm": "^4.0.1"
	}
}
