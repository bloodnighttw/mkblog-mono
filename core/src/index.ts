import path from "node:path";
import {type PluggableList, unified} from "unified";
import remark2rehype from "remark-rehype";
import remarkParse from "remark-parse";
import rehypeStringify from "rehype-stringify";
import fs from "node:fs";
import type {Root as HastRoot} from "hast";
import type {Root as MDRoot, Yaml} from "mdast";
import frontmatter from "remark-frontmatter";
import {select} from "unist-util-select";
import yaml from "yaml";
import {z, type ZodRawShape} from "zod";
import fastGlob, {type Pattern} from "fast-glob";
import {hashing, mix, removeExtension} from "@/utils/slug";

interface ContentOptions<T extends ZodRawShape> {
    /**
     * @param include the glob pattern of the markdown files
     * @param createSlug the function to create the slug of the markdown file,
     * you can use "removeExtension" | "hashing" | "mix" | ((from: string) => string)
     *   - removeExtension: remove the extension of the file
     *   - hashing: hash the file path
     *   - mix: mix the file path and hash
     *   - ((from: string) => string): custom function to create the slug
     *   - default: removeExtension, if not provided
     * @param schema the schema of the metadata, using zod to define and validate the data.
     * @param mkBlogPlugins the function to create the plugins for the markdown blog, not implemented yet.
     * @param remarkPlugins the list of remark plugins
     * @param rehypePlugins the list of rehype plugins
     */
    include: Pattern | Pattern[]
	createSlug?: "removeExtension" | "hashing" | "mix" | ((from: string) => string);
	schema: T;

	mkBlogPlugins?: () => void;
	remarkPlugins?: PluggableList;
	rehypePlugins?: PluggableList;
}

/**
 * @Param T the type of the metadata, we use this to infer the type of the metadata.
 */
class Content<T> {
    /**
     * @param filepath the path of the markdown file
     * @param slug the slug of the markdown file, you can use this to generate the url
     * @param remarkPlugins the list of remark plugins
     * @param rehypePlugins the list of rehype plugins
     * @param schema the schema of the metadata, using zod to define and validate the data.
     * @constructor create a new instance of Content, you should use defineContent to create a collection of Content if
     * you don't know what this is doing.
     * @Return return a new instance of Content
     */
	readonly filepath: string;
	readonly slug: string;
	private readonly remarkPlugins: PluggableList;
	private readonly rehypePlugins: PluggableList;
	private readonly schema: z.ZodSchema<T>;

    constructor(
		filepath: string,
		slug: string,
		remarkPlugins: PluggableList,
		rehypePlugins: PluggableList,
		schema: z.Schema,
	) {
		this.filepath = filepath;
		this.slug = slug;
		this.remarkPlugins = remarkPlugins;
		this.rehypePlugins = rehypePlugins;
		this.schema = schema;
	}

    /**
     * @protected this method should be called internally, it's used to create a remark processor.
     * @Return return a unified processor which can be used to parse markdown content and apply remark plugins.
     */
	protected remarkProcessor() {
		return unified()
			.use(remarkParse)
			.use(frontmatter) // for metadata decode
			.use(this.remarkPlugins);
	}


    /**
     * @protected this method should be called internally, it's used to create a rehype processor.
     * @Return return a unified processor which can be used to convert mdast to hast and apply rehype plugins.
     */
	protected rehypeProcessor() {
		return unified().use(remark2rehype).use(this.rehypePlugins);
	}

    /**
     * @protected this method should be called internally, it's used to create a rehype to html processor.
     * @Return return a unified processor which can be used to convert hast to html string.
     */
	protected rehype2html() {
		return unified().use(rehypeStringify);
	}

    /**
     * @protected this method should be called internally, it's used to parse the markdown file and return the mdast.
     * @Return return the mdast of the markdown file.
     */
	protected async mdast(): Promise<MDRoot> {
		const fileContent = await fs.promises.readFile(this.filepath, "utf-8");
		return this.remarkProcessor().parse(fileContent);
	}


    /**
     * @protected this method should be called internally, it's used to convert the mdast to hast.
     * @param mdroot the mdast root, if it's null, we will fetch mdast automatically
     * @Return return the hast of the markdown file.
     */
	protected async hast(mdroot: MDRoot | null = null): Promise<HastRoot> {
		const temp = mdroot ?? (await this.mdast());
		return await this.rehypeProcessor().run(temp);
	}

	/**
	 * @param root the hast root, if it's null, we will fetch hast automatically
	 * @Return return raw html string of content.
	 */
	public async html(root: HastRoot | null = null) {
		const temp = root ?? (await this.hast());
		return this.rehype2html().stringify(temp);
	}

    /**
     * @param mdast the mdast root, if it's null, we will fetch mdast automatically
     * @Return return the metadata of the markdown file. we use zod to validate the metadata.
     * @throws if metadata not found or metadata is invalid.
     */
	public async metadata(mdast: MDRoot | null = null): Promise<T> {
		const temp = mdast ?? (await this.mdast());
		const yamlContent = select("yaml", temp) as Yaml | null;

		if (yamlContent) {
			const metadata = yaml.parse(yamlContent.value);
			return this.schema.parse(metadata);
		}

		throw new Error("metadata not found");
	}
}

interface Collection<T> {
	posts: Content<T>[];
}

async function defineContent<T extends ZodRawShape>(
	options: ContentOptions<T>,
): Promise<Collection<T>> {

    const files = await fastGlob(options.include);
    const slugFunc =
        options.createSlug === "removeExtension" ? removeExtension :
        options.createSlug === "hashing" ? hashing :
        options.createSlug === "mix" ? mix :
        options.createSlug ?? removeExtension;

	const posts = files.map((file) => {
		const filepath = path.join(process.cwd(), file);
		return new Content<T>(
			filepath,
			slugFunc(file),
			options.remarkPlugins ?? [],
			options.rehypePlugins ?? [],
			z.object(options.schema),
		);
	});

	return { posts };
}

export type { ContentOptions, Collection };

export { defineContent, Content, z as zod };
