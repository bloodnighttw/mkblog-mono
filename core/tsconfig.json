{
	"compilerOptions": {
		"target": "ES2017",
		"module": "esnext",
		"esModuleInterop": true,
		"forceConsistentCasingInFileNames": true,
		"strict": true,
		"skipLibCheck": true,
		"outDir": "dist",
		"moduleResolution": "bundler",
		"paths": {
			"@/*": ["./src/*"],
			"@test/*": ["./test/*"]
		}
	},
	"include": ["src"]
}
