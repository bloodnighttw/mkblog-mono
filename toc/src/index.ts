import {Content, MkblogPlugin} from "@mkblog/core";
import {Element, Root as HRoot} from "hast";
import {h} from "hastscript";
import rehypeSlug from "rehype-slug";
import {visit} from "unist-util-visit";

declare module "@mkblog/core" {
    interface Content<T> {
        rawTOC: () => Promise<HRoot>;
        toc: () => Promise<string>;
    }
}

const tocPlugin: MkblogPlugin = {
    init: () => {
        Content.prototype.rawTOC = async function() {
            const hast = await this.hast();
            const headings: { id: string, text: string, level: number }[] = [];

            visit(hast, 'element', (node: Element) => {
                if (/^h[1-6]$/.test(node.tagName)) {
                    const id = node.properties?.id as string ?? "";
                    const text = node.children
                        .filter(child => child.type === 'text')
                        .map(child => child.value)
                        .join();
                    const level = Number.parseInt(node.tagName[1], 10);
                    headings.push({ id, text, level });
                }
            });

            const buildTOC = (headings: { id: string, text: string, level: number }[], currentLevel = 1): Element => {
                const children: Element[] = [];
                while (headings.length > 0 && headings[0].level >= currentLevel) {

                    const heading = headings.shift()!;
                    const listItem: Element = h('li', { className: `toc-level-${heading.level}` }, [
                        h('a', { href: `#${heading.id}` }, heading.text)
                    ]);

                    if (headings.length > 0 && headings[0].level > heading.level) { // note: null or undefined < 1 === true
                        listItem.children.push(buildTOC(headings, heading.level + 1));
                    }
                    children.push(listItem);
                }
                return h('ul', { className: 'toc' }, children);
            };

            return h(null, buildTOC(headings));
        }

        Content.prototype.toc = async function() {
            const tocTree = await this.rawTOC();
            return this.rehype2html().stringify(tocTree);
        }
    },
    rehypePlugins: [rehypeSlug],
}

export default tocPlugin;