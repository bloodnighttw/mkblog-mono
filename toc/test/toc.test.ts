import {expect, test} from "vitest";
import init from "../src";
import {defineContent,zod} from "../../core/src";
import remarkGfm from "remark-gfm";
import {inspect} from "unist-util-inspect";

test("toc", async ()=> {

    const config = await defineContent({
        include: ["./cases/toc/*.md"],
        schema: {
            title: zod.string(),
        },
        mkBlogPlugins: [
            init
        ],
        remarkPlugins: [remarkGfm],
    })

    const post = config.posts.find((post) => post.slug === "toc");
    // @ts-ignore mono-repo issue
    const tocTree = await post.rawTOC();

    expect(tocTree).toMatchSnapshot();

})