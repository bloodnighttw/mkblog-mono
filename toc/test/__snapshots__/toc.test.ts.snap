// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table of content hast check 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "type": "text",
                  "value": "This is a header",
                },
              ],
              "properties": {
                "href": "#this-is-a-header",
              },
              "tagName": "a",
              "type": "element",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [
                        {
                          "type": "text",
                          "value": "This is a gfm test",
                        },
                      ],
                      "properties": {
                        "href": "#this-is-a-gfm-test",
                      },
                      "tagName": "a",
                      "type": "element",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "type": "text",
                                  "value": "This is a test",
                                },
                              ],
                              "properties": {
                                "href": "#this-is-a-test",
                              },
                              "tagName": "a",
                              "type": "element",
                            },
                          ],
                          "properties": {
                            "className": [
                              "toc-level-3",
                            ],
                          },
                          "tagName": "li",
                          "type": "element",
                        },
                      ],
                      "properties": {
                        "className": [
                          "toc",
                        ],
                      },
                      "tagName": "ul",
                      "type": "element",
                    },
                  ],
                  "properties": {
                    "className": [
                      "toc-level-2",
                    ],
                  },
                  "tagName": "li",
                  "type": "element",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "type": "text",
                          "value": "This is a test",
                        },
                      ],
                      "properties": {
                        "href": "#this-is-a-test-1",
                      },
                      "tagName": "a",
                      "type": "element",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "type": "text",
                                  "value": "This is a test",
                                },
                              ],
                              "properties": {
                                "href": "#this-is-a-test-2",
                              },
                              "tagName": "a",
                              "type": "element",
                            },
                          ],
                          "properties": {
                            "className": [
                              "toc-level-3",
                            ],
                          },
                          "tagName": "li",
                          "type": "element",
                        },
                      ],
                      "properties": {
                        "className": [
                          "toc",
                        ],
                      },
                      "tagName": "ul",
                      "type": "element",
                    },
                  ],
                  "properties": {
                    "className": [
                      "toc-level-2",
                    ],
                  },
                  "tagName": "li",
                  "type": "element",
                },
              ],
              "properties": {
                "className": [
                  "toc",
                ],
              },
              "tagName": "ul",
              "type": "element",
            },
          ],
          "properties": {
            "className": [
              "toc-level-1",
            ],
          },
          "tagName": "li",
          "type": "element",
        },
      ],
      "properties": {
        "className": [
          "toc",
        ],
      },
      "tagName": "ul",
      "type": "element",
    },
  ],
  "type": "root",
}
`;

exports[`Table of content html string check 1`] = `"<ul class="toc"><li class="toc-level-1"><a href="#this-is-a-header">This is a header</a><ul class="toc"><li class="toc-level-2"><a href="#this-is-a-gfm-test">This is a gfm test</a><ul class="toc"><li class="toc-level-3"><a href="#this-is-a-test">This is a test</a></li></ul></li><li class="toc-level-2"><a href="#this-is-a-test-1">This is a test</a><ul class="toc"><li class="toc-level-3"><a href="#this-is-a-test-2">This is a test</a></li></ul></li></ul></li></ul>"`;

exports[`toc 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "type": "text",
                  "value": "This is a header",
                },
              ],
              "properties": {
                "href": "#this-is-a-header",
              },
              "tagName": "a",
              "type": "element",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [
                        {
                          "type": "text",
                          "value": "This is a gfm test",
                        },
                      ],
                      "properties": {
                        "href": "#this-is-a-gfm-test",
                      },
                      "tagName": "a",
                      "type": "element",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "type": "text",
                                  "value": "This is a test",
                                },
                              ],
                              "properties": {
                                "href": "#this-is-a-test",
                              },
                              "tagName": "a",
                              "type": "element",
                            },
                          ],
                          "properties": {
                            "className": [
                              "toc-level-3",
                            ],
                          },
                          "tagName": "li",
                          "type": "element",
                        },
                      ],
                      "properties": {
                        "className": [
                          "toc",
                        ],
                      },
                      "tagName": "ul",
                      "type": "element",
                    },
                  ],
                  "properties": {
                    "className": [
                      "toc-level-2",
                    ],
                  },
                  "tagName": "li",
                  "type": "element",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "type": "text",
                          "value": "This is a test",
                        },
                      ],
                      "properties": {
                        "href": "#this-is-a-test-1",
                      },
                      "tagName": "a",
                      "type": "element",
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "type": "text",
                                  "value": "This is a test",
                                },
                              ],
                              "properties": {
                                "href": "#this-is-a-test-2",
                              },
                              "tagName": "a",
                              "type": "element",
                            },
                          ],
                          "properties": {
                            "className": [
                              "toc-level-3",
                            ],
                          },
                          "tagName": "li",
                          "type": "element",
                        },
                      ],
                      "properties": {
                        "className": [
                          "toc",
                        ],
                      },
                      "tagName": "ul",
                      "type": "element",
                    },
                  ],
                  "properties": {
                    "className": [
                      "toc-level-2",
                    ],
                  },
                  "tagName": "li",
                  "type": "element",
                },
              ],
              "properties": {
                "className": [
                  "toc",
                ],
              },
              "tagName": "ul",
              "type": "element",
            },
          ],
          "properties": {
            "className": [
              "toc-level-1",
            ],
          },
          "tagName": "li",
          "type": "element",
        },
      ],
      "properties": {
        "className": [
          "toc",
        ],
      },
      "tagName": "ul",
      "type": "element",
    },
  ],
  "type": "root",
}
`;
